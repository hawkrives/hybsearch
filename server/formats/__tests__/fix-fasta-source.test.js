/* eslint-env jest */

const { fixFastaSource } = require('../fix-fasta-source')
const { parseFasta } = require('../fasta/parse')
const fs = require('fs')
const path = require('path')

const base = path.join(__dirname, 'input')
const files = fs.readdirSync(base).filter(f => f.endsWith('.fasta'))

for (const file of files) {
	test(file, () => {
		const content = fs.readFileSync(path.join(base, file), 'utf-8')
		const tree = fixFastaSource(content)
		expect(tree).toMatchSnapshot()
	})
}

test('leaves properly-formatted identifiers alone', () => {
	let ident = 'Trionyx_triunguis__HQ012626'
	let input = `>${ident}
-------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCAT`
	let output = fixFastaSource(input)
	expect(parseFasta(output)[0].species).toBe(ident)
})

test('formats an accession+species identifier (where the accession has a period)', () => {
	let input = `>HQ012626.1 Trionyx triunguis voucher Tt24 cytochrome b (cytb) gene, partial cds; mitochondrial
-------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCAT`
	let output = fixFastaSource(input)
	expect(parseFasta(output)[0].species).toBe('Trionyx_triunguis__HQ012626x1')
})

test('formats an accession+species identifier', () => {
	let input = `>HQ012626 Trionyx triunguis voucher Tt24 cytochrome b (cytb) gene, partial cds; mitochondrial
-------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCAT`
	let output = fixFastaSource(input)
	expect(parseFasta(output)[0].species).toBe('Trionyx_triunguis__HQ012626')
})

test('formats an accession-only identifier (with a period)', () => {
	let input = `>HQ012626.1
-------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCAT`
	let output = fixFastaSource(input)
	expect(parseFasta(output)[0].species).toBe('HQ012626x1')
})

test('formats an accession-only identifier', () => {
	let input = `>HQ012626
-------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCAT`
	let output = fixFastaSource(input)
	expect(parseFasta(output)[0].species).toBe('HQ012626')
})
