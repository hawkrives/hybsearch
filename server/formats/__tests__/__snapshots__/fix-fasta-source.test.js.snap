// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emydura-short.fasta 1`] = `
Array [
  Object {
    "sequence": "-ggaacaataaattatcacctcaaaagacacataaaaccacaggaaccccta-ctcaaccataaaaaacacaaatccactattaaaaatcattaacaacacctttattgatctccccaccccatctaacatctctactttatgaaactttggttcattactaggggcatgcctcattctacaactagccacaggaatcttcttagctatacactactcatctgatatttccatagcattctcatcaatctcccacatccaacgagacgttcaatatggctgactaattcgaaatatacacgctaacggtgcttcattattctttatatgcatttacctccatattggacgaggaatctactacggttcatacctttacaagaaaacctgaaacactggagtaatattactactcctagtcatagccactgcattcgtgggctacgtactaccatgaggacaaatatcattctgaggagctacagtaatcaccaatctcctatcagccattccatatgtaggccctacacttgtagaatgaatttgaggggggttctccgtagataacgccacccttacccgattctttacattccacttcctaatcccattcgccatcctaggaataactatactacatctactactactacatgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcctataaagatctactaggccttattctaataataacactcctactcacccttaccctattctaccctaacctactaggagatccagacaactttacaccagccaacccactaaccacc",
    "species": "Emydura_subglobosa__KC755190",
  },
  Object {
    "sequence": "ggaacaatcaattattaccccacaaagacacataaaaccacaggaacccctactccaaccataaaaaacacaaatccactattaaaaatcattaacaacaccttcgttgatcttcccaccccatccaacatctccgctttatgaaacttcggatcattactaggagcatgcctcattctacaactagccacaggaatcttcttagctatacactactcatccgacatttctacagcattctcatcaatttcccacatccaacgagatgttcaatatggatgacttattcgaaatatacacgccaacggtgcttcattattctttatatgcatctaccttcatattggacgagggatctactacggttcctacctttacaacaaaacctgaaacactggagtaatattactactcctagttatagccactgcattcgtgggctacgtgctaccatgaggccaaatatcattctgaggggctacagtaatcaccaacctattatcagccattccatacgtaggcccaacacttgtagagtggatttgaggagggttctccgtagacaacgctactcttactcgattcttcacatttcactttctaatcccattcgccatcctaggaataaccctactacacctcctacttctacacgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcgtataaagacctactaggcctcatcctaataattatatgtctactcacccttaccctatttcacccaaacctactaggagatccagacaactttacaccagccaatccactaaccacc",
    "species": "Emydura_victoriae__KC755189",
  },
  Object {
    "sequence": "ggaacaatcaattattaccccacaaagacacataaaaccacaggaacctctactccaaccataaaaaacacaaatccactattaaaaatcattaacaacaccttcgttgatcttcccaccccatccaacatctccgctttatgaaacttcggatcattactaggagcatgcctcattctacaactagccacaggaatcttcttagctatacactactcatccgacatttctacagcattctcatcaatttcccacatccaacgagatgttcaatatggatgacttattcgaaatatacacgccaacggtgcttcattattctttatatgcatctaccttcatattggacgagggatctactacggttcctacctttacaacaaaacctgaaacactggagtaatattactactcctagttatagccactgcattcgtgggctacgtgctaccatgaggccaaatatcattctgaggggctacagtaatcaccaacctattatcagccattccatacgtaggcccaacacttgtagagtggatttgaggagggttctccgtagacaacgctactcttactcgattcttcacatttcactttctaatcccattcgccatcctaggaataaccctactacacctcctacttctacacgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcgtataaagacctactaggcctcatcctaataattatatgtctactcacccttaccctatttcacccaaacctactaggagatccagacaactttacaccagccaatccactaaccacc",
    "species": "Emydura_victoriae__KC755188",
  },
  Object {
    "sequence": "-ggaacaataa--attatccccaaaagacacataaaactacaggaa-cccctactcaaccataaaaaacacaaacccactactaaaaatcattaataatacctttattgatctccccactccacctaacatctccgctttgtgaaactttggatcattactaggggcatgcctcgctctacaactagccacaggaatcttcttagcgatacactactcatccgatatctccatagcattctcatcaatctcccacatccaacgagatgttcaatatggctgactaattcgaaatatacacgccaacggtgcttcattattctttatatgcatttacctccatattggacgaggaatttactacggttcatatctttacaacaaaacctgaaacactggggtaatattactactcctagtcatagccactgcattcgtgggatacgtactaccatgaggacaaatatcattctgaggagctacagtaatcaccaatctcttatcagccattccatacgtaggccctacacttgtagagtgaatctgagggggattctccgtagatagtgctacccttactcgattcttcacattccactttctaatcccattcgctatcctagggataactatactacatctactactgctacatgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcttataaagatctactaggcctcattctgataatcacacttctactcactcttaccctattctaccccaacctactaggagacccagacaactttacaccagccaacccactaaccacc",
    "species": "Emydura_tanybaraga__KC755187",
  },
  Object {
    "sequence": "-ggaacaataa--attatccccaaaagacacataaaactacaggaa-cccctactcaaccataaaaaacacaaacccactattaaaaatcattaataatacctttattgacctccccactccacctaacatctccgctttgtgaaactttggatcattactaggggcatgcctcgctctacaactagccacaggaatcttcttagcgatacactactcatccgatatctccatagcattctcatcaatctcccacatccaacgagatgttcaatatggctgactaattcgaaatatacacgccaacggtgcttcattattctttatatgcatttacctccatattggacgaggaatttactacggttcatatctttacaacaaaacctgaaacactggggtaatattactactcctagtcatagccactgcattcgtgggatacgtactaccatgaggacaaatatcattctgaggagctacagtaatcaccaatctattatcagccattccatacgtaggccctacacttgtagagtgaatctgagggggattctccgtagatagtgctacccttactcgattcttcacattccactttctaatcccattcgctatcctagggataactatactacatctactattgctacatgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcttataaagatctactaggcctcattctgataatcacacttctactcactcttaccctattctaccccaacctactaggagacccagacaactttacaccagccaacccactaaccacc",
    "species": "Emydura_tanybaraga__KC755186",
  },
  Object {
    "sequence": "-ggaacaataaattaccaccccaaaagacacataaaaccacaggaacccctaccccaaccataaaaaacacaaatccactattaaaaatcattaacaataccttcattgacctccccaccccatctaacatctccgctctatgaaactttgggtcattactaggggcatgccttattctacaactagccacaggaatcttcttagctatacactactcatccgacatttccatagcattctcatcaatctcccacatccaacgagacgttcaatatggctgactaattcgaaatatacacgctaacggtgcttcattattctttatatgcatttacctccatattgggcgaggaatttactacggttcatacctttacaagaaaacttgaaacactggagtaatattactactcctagtaatagccactgcattcgtgggctacgtactaccatgaggacaaatatcattctgaggggctacagtaatcactaatcttctatcagccattccatatgtaggccctacacttgtagaatgaatttgaggaggattctccgtagataacgccacccttacccgattctttacattccactttctaatcccattcgccatcctgggaataactatactacatctactactactacatgaaacaggatcaaataacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcttataaagacctactaggccttattctaataatcacactcctactcacccttaccctgttctaccccaacctactaggagacccagacaactttacaccagccaacccactgaccacc",
    "species": "Emydura_subglobosa__KC755185",
  },
  Object {
    "sequence": "-ggaacaataaattacctccccaaaagacacataaaactacaggaacccctacctcaaccataaaaaacacaaacccactattaaaaatcattaataatacctttattgatctccccactccacctaacatctccgctttgtgaaactttggatcattactaggggcatgcctcgctctacaactagccacaggaatcttcttagcgatacactactcatctgatatctccatagcattctcatcaatctcccacatccaacgagatgttcaatatggctgactaattcgaaatatacacgccaacggtgcttcattattctttatatgcatttacctccatattggacgaggaatttactacggttcatacctttacaacaaaacctgaaacactggggtaatattactactcctagtcatagccactgcattcgtgggatacgtactaccatgaggacaaatatcattctgaggagctacagtaatcaccaacctcttatcagccattccatacgtaggccctacacttgtagagtgaatctgagggggattctccgtagatagtgctacccttactcgattcttcacattccactttctaatcccattcgctatcctagggataactatactacatctactactgctacatgaaacaggatcaaacaacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcttataaagatctactaggcctcattctgataatcacacttctactcactcttaccctattctaccccaacctactaggagacccagacaactttacaccagccaacccactaaccacc",
    "species": "Emydura_subglobosa__KC755184",
  },
  Object {
    "sequence": "-ggaacaatcaattattacccacaaagacacataaaaccacaggaacctctactccaaccataaaaaacacaaatccactattaaaaatcattaataacacctttgttgatcttcccaccccatccaatatctccgccttatgaaacttcggatcattactaggggcatgcctcattctacaactagccacaggaattttcttagccatacactactcatccgacatttctacagcattctcatcaatttcccacatccaacgagatgttcaatatggatgacttattcgaaatatacacgccaacggtgcttcattattctttatatgcatctaccttcatattggacgaggaatctattacggttcctacctttacaataaaacctgaaacactggagtaatattactactactagtcatagccactgcattcgtaggctacgtactaccatgaggccaaatatcattctgaggggctacagtaatcaccaacctattatcagccatcccatacgtaggcccaacacttgtagagtgaatttgaggagggttctccgtagacaacgccacccttactcgattcttcacatttcactttctaatcccattcgccatcctaggaataaccctactacacctcctacttctacacgaaacaggatcaaataacccaacaggactaaactcaaactgtgacaaaatcccattccacccatatttctcgtataaagacctactaggactcatcctaataatcatatgtctacttaccctcaccctattttacccaaacctactaggagatccagacaactttacaccagccaatccactaaccact",
    "species": "Emydura_macquarii__KC755183",
  },
]
`;

exports[`outpluscox1.fasta 1`] = `
Array [
  Object {
    "sequence": "-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCATGAGGGCAAATATCCTTCTGAGG------GGCCACCGTAATCACAAACCTACTATCAGCCGTCCCCTATATCGGCACCACAATAGTACAATGGGTCTGAGGAGGCTTCTCCGTAGACAATGCCA---------CCCTAAC---------------ACGA---------TTCTTCACCATACACTTCCTACTTCCATTCATAATCGCAGGAATAACCATAGT--ACATCTACTATTCCTACACGAAACTGGATCAAACAACCCAATAGGACTTAACTCAAACACAGACAAAATCCC-GTTCCATCCCTACTTCTCATACAAAGACCTACTGGGCCTAACTGTAATACTAACCACACTACTATCTATCGCCATATTTTACCCAAACCTATTGGGGGACCCAGACAACTTCACGCCAGCCAACCCACTATCCACCCCGCCCCACATCAAACCAGAATGATACTTTCTATTCGCATATGCCATCTTACGATCAATCCCAAACAAATTAGGCGGAGTCCTAGCCTTACTACTATCCATCCTAGTGCTATTCACCCTCCCCATAGTCCACACATCAAAACAACGAACACTTACCTTCCGACCAATCACCCAAACCTTATTCTGACTATTCGTAGCCAACCTAATAGTACTGACATGAATTGGAGGTCAACCAGTAGAAAACCCATTCATTCTTATCGGACAAACAGCCTCCATCCTCTATTTCCTAATCTTACTCGTACTAATACC----------------------------------------------------------------------",
    "species": "Trionyx_triunguis__HQ012626x1",
  },
  Object {
    "sequence": "-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ACTAACCATAGCCACCGCATTCATAGGATACGTACTACCATGAGGGCAAATATCCTTCTGAGG------GGCCACCGTAATCACAAACCTACTATCAGCCGTCCCCTATATCGGCACCACAATAGTACAATGGGTCTGAGGAGGCTTCTCCGTAGACAATGCCA---------CCCTAAC---------------ACGA---------TTCTTCACCATACACTTCCTACTTCCATTCATAATCGCAGGAATAACCATAGT--ACATCTACTATTCCTACACGAAACTGGATCAAACAACCCAATAGGACTTAACTCAAACACAGACAAAATCCC-GTTCCATCCCTACTTCTCATACAAAGACCTACTGGGCCTAACTGTAATACTAACCACACTACTATCTATCGCCATATTTTACCCAAACCTATTGGGGGACCCAGACAACTTCACGCCAGCCAACCCACTATCCACCCCGCCCCACATCAAACCAGAATGATACTTTCTATTCGCATATGCCATCTTACGATCAATCCCAAACAAATTAGGCGGAGTCCTAGCCTTACTACTATCCATCCTAGTGCTATTCACCCTCCCCATAGTCCACACATCAAAACAACGAACACTTACCTTCCGACCAATCACCCAAACCTTATTCTGACTATTCGTAGCCAACCTAATAGTACTGACATGAATTGGAGGTCAACCAGTAGAAAACCCATTCATTCTTATCGGACAAACAGCCTCCATCCTCTATTTC---------------------------------------------------------------------------------------------",
    "species": "Trionyx_triunguis__HQ012625x1",
  },
]
`;
